### PR TITLE
MTD simulation: update Track ID stored in MTD SimHits

### PR DESCRIPTION
**Pull request by fabiocos**
#### PR description:

This PR addresses two related issues, emerged in the past both in MTD studies and in discussions with HGCal experts (@rovere @felicepantaleo):

- provide the MC truth for BTL clusters, intended as collection of adjacent crystals within a module hit by the same track entering from tracker;
- provide a working association between ETL hits and simulated tracks stored in the event, like it is happening for any calorimetric hit.

The latter is simply solved by instrumenting the ```MtdSD``` class with code using the ```caloIDonSurface``` mechanism implemented in ```TrackInformation```.
The former requires to implement a mechanism similar to that used for the tracker -> calo transition, with the additional complication that BTL is fully embedded in the tracker, and this can generate a bit more complex set of particles configurations.

In order to address this last issue, the ```TrackInformation``` class is instrumented with MTD-specific information, packed into a bitted word to avoid multiple ```bool``` variables and save memory space. The information about transitions Tracker -> BTL and BTL -> Tracker are managed within the ```SteppingAction``` class, while the handling of tracks created within the BTL volume is made in the ```NewTrackAction``` class.

The present PR is not supposed to modify the simulation history in terms of track / vertices stored in the event, or hit collections. Only the track ID stored within BTL/ETL hits can be affected. This PR is not completely addressing the problem of tracks created within the BTL volume, at present not stored in the event persistent history, and then entering the calo volume. The cut mechanism enabling the persistent storage is implemented in ```MtdSD```, but left switched off by default (using the standard cuts effectively preventing any action). The possible update of this part, affecting the event history, is deferred to a separate PR.

This PR is in principle ineffective on run2/run3 simulation, but some checks in ```SteppingAction``` and ```NewTrackAction``` are running also for that. Without an era-dependent flag, I do not see a way to avoid it.

The update of existing MTD geometry regions for ProdCuts should be totally transparent, as they are so far used in one and only one place within CMSSW, and the existing behaviour is kept. 

For further discussion, see https://indico.cern.ch/event/1275985/contributions/5359348/attachments/2628038/4545909/MTD-SIM_20230414.pdf

#### PR validation:

The code behaviour has been studies through detailed history and debugging printout on single muon, electron and pion guns, verifying the correctness of the logic for various cases: particle entering BTL, exiting BTL, loopers re-entering BTL, particles crossing ETL. A test on run2 simulation test wf 11607.0 do not show any difference, as expected.

### CODE REVIEW

**Comment by fabiocos** [comment]
@civanch we need to understand how to manage this PR wrt to yours #41315

**Comment by fabiocos** [comment]
please test workflow 20834.0

this needs to work also for D88 (i.e. old BTL geometry), as soon as it is supported

**Comment by civanch** [comment]
@fabiocos , sorry for the mess - I prepared my PR during holiday, likely you did the same. As we discussed before about MC truth, there is no principal contradictions between these PRs, only technical. As far as I see two places:
   1) I have updated TrackInformation class making constructor public and initialize class members in a different way;
   2) substituted NewTrackAction by MCTruthUtil, so it is not anymore class by namespace with utility methods; extra utility method may be added here.

What may be a problem (also we discussed before): I would prefer if SD classes fill TrackInformation container and not fill or create TrackWIthHistory objects. As we agreed, we should first discuss this approach with all involved parties and only after make a migration. The main reason - TrackInformation is design to store this information, we cannot store it in two places. The best approach is to extend TrackInformation if needed and not fill part of information there, another part - in TrackWithHistory with badly controlled relation between two.

**Comment by fabiocos** [comment]
@civanch as you will see reviewing the code, I am not touching ```TrackWithHistory```, all my update is based on ```TrackInformation```.  Just let me know whether you would prefer to go first with your update, I will need to rebase my code on top of it.

**Comment by civanch** [comment]
@fabiocos , as usual when rename something some hidden problem appears. Our both PRs required fixes. I start to think that optimal would be to merge yours first, after mine. So, please, fix compilation in this PR and let us push forward.

**Comment by fabiocos** [comment]
The problem appears in the reconstruction geometry, where, contrary to my statement, the ```FastTimeRegion``` label is used, in https://github.com/cms-sw/cmssw/blob/master/Geometry/MTDNumberingBuilder/src/CmsMTDStringToEnum.cc#L6 . Solution to be provided.

**Comment by fabiocos** [comment]
I had unfortunately missed that rhe ```MTDNumberingBuilder``` construction uses, for DDD only, cms cut regions, see https://github.com/cms-sw/cmssw/blob/master/Geometry/MTDNumberingBuilder/plugins/DDCmsMTDConstruction.cc#L48 , instead of other attributes, as better done by the DD4hep version of the code, see https://github.com/cms-sw/cmssw/blob/master/Geometry/MTDNumberingBuilder/plugins/DDCmsMTDConstruction.cc#L273 . 
In order to minimize as much as possible changes for DDD, that will be hopefully abandoned sooner or later, I updated the map of allowed names used for regions, and it appears to work correctly. Geometry unit tests are ok for both D88 and D95 (DDD vs DD4hep, no PR vs this PR), and step2 runs.

Looking for instances of ```FastTimerRegion``` in both ```*.h``` and ```*.cc``` files in the whole CMSSW repository does not show other instances.

**Comment by fabiocos** [comment]
please test workflow 20834.0

**Comment by fabiocos** [comment]
The output of the relval tests appears at least qualitatively consistent with expectations:

- only MTD histograms are affected
- SimHits histogram changes show a reduction of the number of track ID associated with energy deposits in MTD cells, consistent with the fact that less track IDs are now stored in the MTD hits, since several of them are moved to the parent track;
- tracking histograms are consistent with an increase of TrackingParticles with a G4Track appearing in a MTD sim hit, and there are less DetIds with unassociated track IDs in sim hits;
- no further change is observed in simulated or reconstructed quantities.

**Comment by civanch** [comment]
+1

File: SimG4CMS/Forward/src/MtdSD.cc [hunk 1]
```diff
@@ -40,6 +41,11 @@ MtdSD::MtdSD(const std::string& name,
   if (scheme)
     setNumberingScheme(scheme);
 
+  energyCut = m_p.getParameter<double>("EnergyThresholdForPersistencyInGeV") * CLHEP::GeV;     //default must be 0.5
+  energyHistoryCut = m_p.getParameter<double>("EnergyThresholdForHistoryInGeV") * CLHEP::GeV;  //default must be 0.05
+
+  setCuts(energyCut, energyHistoryCut);
```
Comment (fabiocos) [inline SimG4CMS/Forward/src/MtdSD.cc:47]: this statement would treat tracks in BTL and in ETL in the same way, which is probably something we do not want. I suggest we possibly discuss this for a further development.

Comment (civanch) [reply SimG4CMS/Forward/src/MtdSD.cc:47]: @fabiocos , yes, in the next rounds I would prefer remove these lines from all SD classes - it is not their job. For now this PR is fine. @srimanob, would you agree?

**Comment by civanch** [comment]
@srimanob, @AdrianoDee, please, sign this PR, we need it to be merged in order to continue activity on MC truth.

**Comment by srimanob** [comment]
+Upgrade

From the upgrade-related package, the update on `Geometry/MTD*` is on existing MTD geometry regions for ProdCuts and MTDNumberingBuilder. The PR test looks fine, and only Phase-2 MTD histograms are effected.

**Pull request merged**